File: app/src/main/java/com/melvinhou/dimension2/MainActivity.java
package com.melvinhou.dimension2;

import android.annotation.SuppressLint;
import android.text.TextUtils;
import android.view.Gravity;
import android.view.MenuItem;

import com.melvinhou.dimension2.databinding.ActMainBD;
import com.melvinhou.kami.mvvm.DataBindingActivity;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

import androidx.appcompat.graphics.drawable.DrawerArrowDrawable;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import androidx.navigation.ui.AppBarConfiguration;
import androidx.navigation.ui.NavigationUI;


/**
 * ===============================================
 * = 作 者：风 尘
 * <p>
 * = 版 权 所 有：melvinhou@163.com
 * <p>
 * = 地 点：中 国 北 京 市 朝 阳 区
 * <p>
 * = 时 间：2020/7/1 0:0
 * <p>
 * = 分 类 说 明：主页
 * ================================================
 */
public class MainActivity extends DataBindingActivity<ActMainBD> {

    private AppBarConfiguration mAppBarConfiguration;
    private NavController mNavController;

    @Override
    protected int getLayoutID() {
        return R.layout.activity_main;
    }

    @Override
    protected void initView() {
        setSupportActionBar(getViewDataBinding().bar);
        //显示自带title
        getSupportActionBar().setDisplayShowTitleEnabled(true);

        DrawerArrowDrawable drawable= new DrawerArrowDrawable(
                getDelegate().getSupportActionBar().getThemedContext());
//        getSupportActionBar().setIcon(drawable);
        getSupportActionBar().setHomeAsUpIndicator(drawable);
    }

    @Override
    protected void initListener() {
        //绑定
        mNavController = Navigation.findNavController(this, R.id.nav_host_fragment);
        //id用于匹配当前点击位置
        AppBarConfiguration appBarConfiguration = new AppBarConfiguration.Builder(
                R.id.navigation_home, R.id.navigation_media, R.id.navigation_function)
                .build();
        mAppBarConfiguration = new AppBarConfiguration.Builder(
                R.id.navigation_home, R.id.navigation_gallery, R.id.navigation_slideshow)
                .setDrawerLayout(getViewDataBinding().drawerLayout)
                .build();
//        NavigationUI.setupActionBarWithNavController(this, mNavController, appBarConfiguration);
        //为了不执行AbstractAppBarOnDestinationChangedListener的setNavigationIcon
        mNavController.addOnDestinationChangedListener((controller, destination, arguments) -> {
            CharSequence label = destination.getLabel();
            if (!TextUtils.isEmpty(label)) {
                // Fill in the data pattern with the args to build a valid URI
                StringBuffer title = new StringBuffer();
                Pattern fillInPattern = Pattern.compile("\\{(.+?)\\}");
                Matcher matcher = fillInPattern.matcher(label);
                while (matcher.find()) {
                    String argName = matcher.group(1);
                    if (arguments != null && arguments.containsKey(argName)) {
                        matcher.appendReplacement(title, "");
                        //noinspection ConstantConditions
                        title.append(arguments.get(argName).toString());
                    } else {
                        throw new IllegalArgumentException("Could not find " + argName + " in "
                                + arguments + " to fill label " + label);
                    }
                }
                matcher.appendTail(title);
                setTitle(title);
            }

            /*
            boolean isTopLevelDestination = NavigationUI.matchDestinations(destination,
                    appBarConfiguration.getTopLevelDestinations());
            if (drawerLayout == null && isTopLevelDestination) {
                setNavigationIcon(null, 0);
            } else {
                setActionBarUpIndicator(drawerLayout != null && isTopLevelDestination);
            }
            */
        });
        NavigationUI.setupWithNavController(getViewDataBinding().btnNavView, mNavController);
//        NavigationUI.setupActionBarWithNavController(this, mNavController, mAppBarConfiguration);
        NavigationUI.setupWithNavController(getViewDataBinding().navView, mNavController);

    }

    @Override
    protected void initData() {
    }


    @Override
    public boolean onSupportNavigateUp() {
//        return NavigationUI.navigateUp(mNavController, mAppBarConfiguration)
//                || super.onSupportNavigateUp();
        return super.onSupportNavigateUp();
    }

    @SuppressLint("WrongConstant")
    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                getViewDataBinding().drawerLayout.openDrawer(Gravity.START, true);
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onLoading() {
//        super.onLoading();

    }

}

File: app/src/main/java/com/melvinhou/dimension2/MainActivity.kt
package com.melvinhou.dimension2

import android.view.Gravity
import android.view.MenuItem
import androidx.appcompat.graphics.drawable.DrawerArrowDrawable
import androidx.core.view.WindowCompat
import androidx.navigation.NavController
import androidx.navigation.findNavController
import androidx.navigation.ui.AppBarConfiguration
import androidx.navigation.ui.setupActionBarWithNavController
import androidx.navigation.ui.setupWithNavController
import com.google.android.material.bottomnavigation.BottomNavigationView
import com.melvinhou.dimension2.databinding.ActivityMainBinding
import com.melvinhou.kami.mvvm.BaseViewModel
import com.melvinhou.kami.mvvm.BindActivity


/**
 * ===============================================
 * = 作 者：风 尘
 * <p>
 * = 版 权 所 有：melvinhou@163.com
 * <p>
 * = 地 点：中 国 北 京 市 朝 阳 区
 * <p>
 * = 时 间：2023/3/17 0017 9:37
 * <p>
 * = 分 类 说 明：
 * ================================================
 */
class MainActivity : BindActivity<ActivityMainBinding, BaseViewModel>() {
    override fun openViewBinding(): ActivityMainBinding =
        ActivityMainBinding.inflate(layoutInflater)

    override fun openModelClazz(): Class<BaseViewModel> =
        BaseViewModel::class.java


    private lateinit var mNavController: NavController
    private lateinit var mNavView: BottomNavigationView

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.home -> {
                mBinding.drawerLayout.openDrawer(Gravity.LEFT, true)
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun initView() {
        //添加一个toolbar，阻止底部导航栏setTitle报错
        setSupportActionBar(mBinding.bar)
        //显示标题
        supportActionBar!!.setDisplayShowTitleEnabled(true)
        val drawable = DrawerArrowDrawable(baseContext)
        supportActionBar?.setHomeAsUpIndicator(drawable);
        supportActionBar?.setDisplayShowHomeEnabled(true)

        //
        mNavController = findNavController(R.id.nav_host_fragment)
        mNavView = mBinding.btnNavView
        val appBarConfiguration = AppBarConfiguration(
            setOf(
                R.id.navigation_home,
                R.id.navigation_media,
                R.id.navigation_function
            )
        )
        setupActionBarWithNavController(mNavController, appBarConfiguration)
        mNavView.setupWithNavController(mNavController)
        mBinding.navView.setupWithNavController(mNavController)
        //切换时保留原始色彩
//        mNavView.itemIconTintList = null
    }


    override fun initListener() {
        //自定义跳转
        mNavView.setOnItemSelectedListener {
            mNavController.navigate(it.itemId)
            true
        }
        //监听，切换状态栏颜色
        mNavController.addOnDestinationChangedListener { _, destination, _ ->

            val isblank = when (destination.id) {
//                R.id.navigation_home -> false
                else -> true
            }
            WindowCompat.getInsetsController(window, mBinding.root).let {
                it.isAppearanceLightStatusBars = isblank
                it.isAppearanceLightNavigationBars = true
            }
        }
    }

    override fun initData() {

    }

}